src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -180,47 +179,306 @@ protected override void HandlerRemovedInternal(IChannelHandlerContext context)
 
         protected override void Decode(IChannelHandlerContext context, IByteBuffer input, List<object> output)
         {
-            // pass bytes to SslStream through input -> trigger HandleSslStreamRead. After this call sslStreamReadBuffer may or may not have bytes to read
-            this.mediationStream.AcceptBytes(input);
+            int startOffset = input.ReaderIndex;
+            int endOffset = input.WriterIndex;
+            int offset = startOffset;
+            int totalLength = 0;
 
-            if (!this.EnsureAuthenticated())
+            List<int> packetLengths;
+            // if we calculated the length of the current SSL record before, use that information.
+            if (this.packetLength > 0)
             {
-                return;
+                if (endOffset - startOffset < this.packetLength)
+                {
+                    return;
+                }
+                else
+                {
+                    packetLengths = new List<int>(4);
+                    packetLengths.Add(this.packetLength);
+                    offset += this.packetLength;
+                    totalLength = this.packetLength;
+                    this.packetLength = 0;
+                }
+            }
+            else
+            {
+                packetLengths = new List<int>(4);
+            }
+
+            bool nonSslRecord = false;
+
+            while (totalLength < TlsUtils.MAX_ENCRYPTED_PACKET_LENGTH)
+            {
+                int readableBytes = endOffset - offset;
+                if (readableBytes < TlsUtils.SSL_RECORD_HEADER_LENGTH)
+                {
+                    break;
+                }
+
+                int packetLength = TlsUtils.GetEncryptedPacketLength(input, offset);
+                if (packetLength == -1)
+                {
+                    nonSslRecord = true;
+                    break;
+                }
+
+                Contract.Assert(packetLength > 0);
+
+                if (packetLength > readableBytes)
+                {
+                    // wait until the whole packet can be read
+                    this.packetLength = packetLength;
+                    break;
+                }
+
+                int newTotalLength = totalLength + packetLength;
```

Comment:
btw i assume this also includes the bytes involved in encoding the length?

the comment for GetEncryptedPacketLength says that the reader index is not incremented